### PR TITLE
Generic support for uni gmm

### PR DESCRIPTION
**Pull request by xukai92**
Add generic support for uni-gmm

### CODE REVIEW

**Comment by ararslan** [comment]
These are now parametric types, you'll have to modify the constructors to account for the type parameter. The 0.5-compatible syntax is
```julia
function (::Type{UnivariateGMM{T}}){T}(...)
   ...
end
```
(Much nicer syntax for this is available in 0.6 but unfortunately it isn't compatible with 0.5.)

**Comment by xukai92** [comment]
Thank you for pointing this. I will make the change soon. I have a problem which I don't know if I can ask here - actually I failed to run the tests for `Distributions.jl` on my local, with error shown below

```julia
Pkg.test("Distributions")
INFO: Computing test dependencies for Distributions...
ERROR: unsatisfiable package requirements detected: no feasible version could be found for package: ColorVectorSpace
```

Do you know why this happens?

**Comment by ararslan** [comment]
I've seen that error once before with a different package. If `Pkg.update()` shows unexpected warnings, you can always `cd ~/.julia/v0.x/METADATA && git fetch origin && git reset --hard origin/metadata-v2` (where 0.x is your current Julia version, e.g. 0.6). If none of that works, I've heard that a solution that works is to delete the contents of `~/.julia/v0.x`...

**Comment by xukai92** [comment]
Hi Alex, I fix the constructor now. 

I stoped working on this PR because I also meet some dependent problems - a group of functions used by uni gmm seems didn't support `Dual` **before**, and I didn't manage to fix all of them. But today after I fix the constructor and re-run the them, and the tests seems pass on my local now.

File: src/mixtures/unigmm.jl [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
+immutable UnivariateGMM{T<:Real} <: UnivariateMixture{Continuous,Normal}
     K::Int
-    means::Vector{Float64}
-    stds::Vector{Float64}
+    means::Vector{T}
+    stds::Vector{T}
     prior::Categorical
 
-    function UnivariateGMM(ms::Vector{Float64}, ss::Vector{Float64}, pri::Categorical)
+    (::Type{UnivariateGMM{T}}){T}(ms::Vector{T}, ss::Vector{T}, pri::Categorical) = begin
```
Comment (ararslan) [inline src/mixtures/unigmm.jl]: Why change from the long-form `function` definition? That's generally preferable to `begin` blocks for style.

File: src/mixtures/unigmm.jl [hunk 2]
```diff
@@ -1,20 +1,22 @@
...
         K = length(ms)
         length(ss) == K || throw(DimensionMismatch())
         ncategories(pri) == K ||
             error("The number of categories in pri should be equal to the number of components.")
-        new(K, ms, ss, pri)
+        new{T}(K, ms, ss, pri)
     end
 end
 
+UnivariateGMM{T<:Real}(ms::Vector{T}, ss::Vector{T}, pri::Categorical) = UnivariateGMM{T}(K, ms, ss, pri)
```
Comment (ararslan) [inline src/mixtures/unigmm.jl]: `K` isn't defined here

**Comment by xukai92** [comment]
Thx Alex. I found that I messed up 3 Distributions.jl forks on my local so I actually run tests on another commit. I will give another try to fix it.

**Comment by xukai92** [comment]
Ok, I think I meet the problem I see before: a lot of functions in `mixturemodel.jl` are not generic. 
- `cdf` related functions sees not generic support
  - I would like to skip these first and do `pdf` related frist
- I separated the test for generic unigmm and commented our `cdf` related functions now
  - I tried to make some `pdf` related functions in `mixturemodel.jl` and now I face some numerical overflow?

```julia
julia test/mixture.jl
    testing UnivariateMixture
Test Failed
  Expression: logpdf(g,X[i]) ≈ mix_lp0[i]
   Evaluated: -Inf isapprox -2.0463780454938103
ERROR: LoadError: There was an error during testing
 in record(::Base.Test.FallbackTestSet, ::Base.Test.Fail) at test.jl:397
 in do_test(::Base.Test.Returned, ::Expr) at test.jl:281
 in test_mixture(::Distributions.MixtureModel{Distributions.Univariate,Distributions.Continuous,Distributions.Normal}, ::Int64, ::Int64) at mixture.jl:56
 in include_from_node1(::String) at loading.jl:488
 in include_from_node1(::String) at sys.dylib:?
 in process_options(::Base.JLOptions) at client.jl:265
 in _start() at client.jl:321
 in _start() at sys.dylib:?
while loading /Users/kai/projects/Turing/exps/Distributions.jl/test/mixture.jl, in expression starting on line 212
```

Can you give me some help and some general suggestions how to make things in `mixturemodel.jl` generics

**Comment by matbesancon** [comment]
Hi @xukai92 what's the status on this? Maybe you can keep going on this PR on Julia 1.0 and the latest Distributions master branch

**Comment by matbesancon** [comment]
ping @xukai92 to move on this :)

**Comment by xukai92** [comment]
@matbesancon Thanks for pinging me! I'm on it now :)

**Comment by xukai92** [comment]
I disabled the tests for generic support of GMM for `rand(g, n)` as this is not supported for `Normal`, i.e. below fails:
```julia
julia> d = Normal(Dual(0), Dual(1))
Normal{Dual{Nothing,Int64,0}}(μ=Dual{Nothing}(0), σ=Dual{Nothing}(1))

julia> rand(d, 10)
ERROR: MethodError: no method matching Float64(::Dual{Nothing,Float64,0})
Closest candidates are:
  Float64(::Real, ::RoundingMode) where T<:AbstractFloat at rounding.jl:194
  Float64(::T<:Number) where T<:Number at boot.jl:741
  Float64(::Int8) at float.jl:60
  ...
Stacktrace:
 [1] convert(::Type{Float64}, ::Dual{Nothing,Float64,0}) at ./number.jl:7
 [2] setindex!(::Array{Float64,1}, ::Dual{Nothing,Float64,0}, ::Int64) at ./array.jl:767
 [3] rand! at /Users/kai/projects/Distributions.jl/src/univariates.jl:167 [inlined]
 [4] rand at /Users/kai/projects/Distributions.jl/src/univariates.jl:160 [inlined]
 [5] rand(::Normal{Dual{Nothing,Int64,0}}, ::Int64) at /Users/kai/projects/Distributions.jl/src/genericrand.jl:24
 [6] top-level scope at none:0
```

**Comment by xukai92** [comment]
@matbesancon Not sure why some tests for `Hypergeometric` fail.

**Comment by matbesancon** [comment]
yes sadly we need #836 merged for this first

**Comment by matbesancon** [comment]
not sure if we should do it here or in a following PR, but it would be good to replace `Vector` by a new implementation

**Comment by xukai92** [comment]
Do yo mean making everything typed by `AT, T` where `AT<:AbstractVector{T}, T<:Real`?

**Comment by matbesancon** [comment]
yes exactly

**Comment by matbesancon** [comment]
Also what was done in #837 for multivariate gaussians

**Comment by matbesancon** [comment]
ping @xukai92 all good on this?

File: src/mixtures/mixturemodel.jl [hunk 3]
```diff
@@ -135,13 +135,13 @@ function MixtureModel(components::Vector{C}, prior::Categorical) where C<:Distri
     MixtureModel{VF,VS,C}(components, prior)
 end
 
-MixtureModel(components::Vector{C}, p::Vector{Float64}) where {C<:Distribution} =
+MixtureModel(components::Vector{C}, p::Vector{T}) where {C<:Distribution,T<:Real} =
```
Comment (matbesancon) [inline src/mixtures/mixturemodel.jl]: Should we allow for any `<:AbstractVector{<:Real}`?

Comment (xukai92) [reply src/mixtures/mixturemodel.jl]: Is this the only place we need to change? Or we actually need make `AT<: AbstractVector` part of the parameter of type `MixtureModel`?

Comment (matbesancon) [reply src/mixtures/mixturemodel.jl]: yes exactly

File: src/mixtures/unigmm.jl [hunk 4]
```diff
@@ -1,20 +1,22 @@
 # Univariate Gaussian Mixture Models
 
-struct UnivariateGMM <: UnivariateMixture{Continuous,Normal}
+struct UnivariateGMM{T<:Real} <: UnivariateMixture{Continuous,Normal}
```
Comment (matbesancon) [inline src/mixtures/unigmm.jl]:
```suggestion
struct UnivariateGMM{VT1<:AbstractVector{<:Real},VT2<:AbstractVector{<:Real}} <: UnivariateMixture{Continuous,Normal}
means::VT1
stds::VT2
```

This also allows for mean and std to have different element types (see the discussion in https://github.com/JuliaStats/Distributions.jl/pull/837 for motivation)

File: src/mixtures/unigmm.jl [hunk 5]
```diff
@@ -1,20 +1,22 @@
...
     K::Int
-    means::Vector{Float64}
-    stds::Vector{Float64}
+    means::Vector{T}
+    stds::Vector{T}
     prior::Categorical
 
-    function UnivariateGMM(ms::Vector{Float64}, ss::Vector{Float64}, pri::Categorical)
+    function UnivariateGMM{T}(ms::Vector{T}, ss::Vector{T}, pri::Categorical) where {T<:Real}
```
Comment (matbesancon) [inline src/mixtures/unigmm.jl]: Don't think we need the type parameter in the constructor `UnivariateGMM{T}`, given that it's also in the `where` clause

File: src/mixtures/unigmm.jl [hunk 6]
```diff
@@ -1,20 +1,22 @@
...
         K = length(ms)
         length(ss) == K || throw(DimensionMismatch())
         ncategories(pri) == K ||
             error("The number of categories in pri should be equal to the number of components.")
-        new(K, ms, ss, pri)
+        new{T}(K, ms, ss, pri)
     end
 end
 
+UnivariateGMM(ms::Vector{T}, ss::Vector{T}, pri::Categorical) where {T<:Real} = UnivariateGMM{T}(ms, ss, pri)
```
Comment (matbesancon) [inline src/mixtures/unigmm.jl]: this constructor becomes redundant given previous comment

**Comment by xukai92** [comment]
@matbesancon suggestions fixed! Please take another review.

**Comment by matbesancon** [comment]
Approved, merged in 12 hours if no objection

**Comment by xukai92** [comment]
Thanks for the help @matbesancon !

**Comment by matbesancon** [comment]
woops, 12 hours became 2 days, merged. Thanks for the contribution @xukai92

**Pull request merged**